ktor-server/ktor-server-plugins/ktor-server-rate-limit/jvmAndNix/src/io/ktor/server/plugins/ratelimit/RateLimit.kt

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.server.plugins.ratelimit
+
+import io.ktor.http.*
+import io.ktor.server.application.*
+import io.ktor.server.response.*
+import io.ktor.server.routing.*
+import io.ktor.util.*
+import io.ktor.util.collections.*
+import io.ktor.util.date.*
+import io.ktor.util.logging.*
+import kotlin.time.*
+
+internal val LOGGER = KtorSimpleLogger("io.ktor.server.plugins.RateLimit")
+private const val LIMITER_NAME_GLOBAL = "KTOR_GLOBAL_RATE_LIMITER"
+internal const val LIMITER_NAME_EMPTY = "KTOR_NO_NAME_RATE_LIMITER"
+
+internal val RateLimiterInstancesRegistryKey =
+    AttributeKey<ConcurrentMap<Pair<String, Any>, RateLimiter>>("RateLimiterInstancesRegistryKey")
+
+internal val RateLimiterConfigsRegistryKey =
+    AttributeKey<Map<String, RateLimitProvider>>("RateLimiterConfigsRegistryKey")
+
+/**
+ * A plugin that provides rate limiting for incoming requests.
+ */
+public val RateLimit: ApplicationPlugin<RateLimitConfig> = createApplicationPlugin("RateLimit", ::RateLimitConfig) {
```

Comment:
Maybe we should add config file support

Reply:
we can add it only for the most basic cases, because more advanced requires lambda functions